From d2e4d015a3a387220873b5c2196b047f03ae17b6 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Tue, 22 Nov 2016 14:09:12 +0200
Subject: [PATCH 0597/1345] fix: net: mvpp2x: DMA unmap buffers during BM free
 procedure

commit  427c3483e9a983bf9e8706d382c5ea53b073dd17 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- Buffers should be unmapped during BM free procedure since
  buffers were DMA mapped during mv_pp2x_rx_refill_new.
  Fix memory leakage during BM free procedure.

Change-Id: Ia9acbcd9e1f7c4e6a3319504b40521bb6af0a731
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/33957
Reviewed-by: Yuval Caduri <cyuval@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
index 6a7a529..80664aa 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
@@ -385,6 +385,8 @@ void mv_pp2x_bm_bufs_free(struct device *dev, struct mv_pp2x *priv,
 		if (!phys_addr)
 			break;
 		if (!bm_pool->external_pool) {
+			dma_unmap_single(dev, phys_addr,
+				 MVPP2_RX_BUF_SIZE(bm_pool->pkt_size), DMA_TO_DEVICE);
 			virt_addr = phys_to_virt(dma_to_phys(dev, phys_addr));
 			mv_pp2x_frag_free(bm_pool, virt_addr);
 		}
-- 
1.7.9.5

